Improve multi-screen support

etiktin opened the pull request:
Resolves #1.

The first issue was solved by keeping track of the display bounds. If the display is not available when we load, we'll drop the state and use the defaults.

The second issue, was solved by listening to the `move` and `resize` events and updating the state. We're using timers to eliminate concerns of querying Electron's state too often. And it's all wrapped in a new method called `register` that takes care of everything automatically, including saving the state on `close`.

This was tested on Windows, but should work fine on OS X and Linux.

mawie81 commented:
Thank you for the PR!

I just checked this out on a Mac. Overall it´s looking really good.

I somehow (not sure how yet) ran into the issue of having no x and y coordinates in the saved json file. ```getDisplayMatching```does not like this on the next app start. One way to avoid this is to always save the window position and size in ```saveState``` even if it is fullscreen, minimized or maximized - haven´t tested Windows yet though. What do you think?

And as another idea: how about moving the call for ```win.maximize()```out of the app using the module and inside the current ```register``` function (maybe it needs another name then)? Then users of the module can just call ```register``` and don´t need additional boilerplate.
This way the also still missing part of storing the isFullScreen flag and its corresponding ```setFullScreen(...)``` call could be added there as well.

etiktin commented:
Thanks for your feedback :)

I fixed the issue with the missing coordinates, by making the window bounds mandatory, so if any are missing we'll just drop the state (instead of crashing). The display bounds are not mandatory, so users of older versions can update without losing their previous state on the first run. You can test it by trying to delete parts of the JSON and loading the app.
Also tried to simplify some of the code (e.g. unified the `move` and `resize` handlers, added a `updateState` function so that logic won't be duplicated any more etc.).

I like your idea about maximizing the window for the user in the `register` function. It will make usage even simpler. I think we should probably add an option to disable this behavior, so the user could control it. I'm suggesting this since some apps on Windows (e.g. Notepad) keep position and size correctly, but don't start as maximized if you closed it as maximized, so I guess some users might want to do the same.
In regards to names, what do you think about the following:
- `register` => `manage` (e.g. `winStateManager.manage(win)`)
- `unregister` => `unmanage`
- `dontMaximize` as a boolean option for disabling auto maximizing

In regards to fullscreen, at least on Windows, you can't manually move it or resize it and if you minimize the window `win.isFullScreen` still returns true. So it's a bit weird and I'm not sure what's the benefit of using `electron-window-state` with it (since the user can't move or resize the window). 
Does it work differently on OS X or Linux? 
In any case, we can probably postpone adding support for it to another PR.

mawie81 commented:
Ok gave it a test on Mac again. All looking fine for now :)

I´m fine with naming ```register``` to ```manage``` and ```unregister```to ```unmanage``` so the window can  be maximized in there. A flag to control this would be good too so that a user could opt out if needed.

One last thing: why do we need ```unregister``` in the public API? Shouldn´t it be enough to call it internally when the window is closing? Not sure why someone would want to opt out before.

About the fullscreen handling: as a user I would expect that the app gets restored to fullscreen when I closed it in this state. But I agree that this should not be in the scope of this PR.

Could you do the renaming and maybe (if possible) remove the ```unregister``` from the public API?
After that and some more Windows tests I´m happy to merge :)

etiktin commented:
I removed `unregister` from the exposed API, changed `register` => `manage` and added an option to opt out. Also ended up adding support for full screen (your argument makes sense).

What do you think? Can we :ship: it?

mawie81 commented:
Landed :)
Thank you very much for the time and effort you put into this.

etiktin commented:
Awesome, thanks :)
You're welcome. It was fun working on it.
